src/main/java/org/elasticsearch/plugins/repository/LocalPluginRepository.java

```diff
@@ -0,0 +1,433 @@
+import org.elasticsearch.plugins.PluginsService;
+
+import java.io.IOException;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.nio.file.attribute.PosixFileAttributeView;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.Collection;
+import java.util.Iterator;
+
+import static org.elasticsearch.plugins.repository.PluginDescriptor.Builder.pluginDescriptor;
+
+/**
+ * Local plugin repository used to manage installed plugins.
+ */
+public class LocalPluginRepository implements PluginRepository {
+
+    public static final ImmutableSet<String> FORBIDDEN_NAMES = ImmutableSet.of("elasticsearch", "elasticsearch.bat", "elasticsearch.in.sh",
+                                                                                "plugin", "plugin.bat", "service.bat");
+
+    private final static String BIN = "bin";
+    private final static String CONFIG = "config";
+    private final static String SITE = "_site";
+    private final static String JAR = ".jar";
+    private final static String CLASS = ".class";
+    private final static String JAVA = ".java";
+
+    private final String name;
+
+    private final Path pluginsDir;
+    private final Path binDir;
+    private final Path configDir;
+
+    private final Settings settings;
+
+    private LocalPluginRepository(String name, Environment env, Settings settings) {
+        this.name = name;
+
+        this.binDir = env.homeFile().resolve(BIN);
+        this.configDir = env.configFile();
+        this.pluginsDir = env.pluginsFile();
+
+        this.settings = settings;
+    }
+
+    @Override
+    public String name() {
+        return name;
+    }
+
+    @Override
+    public Collection<PluginDescriptor> find(final String name) {
+        return list(new Filter() {
+            @Override
+            public boolean accept(PluginDescriptor plugin) {
+                return StringUtils.equalsIgnoreCase(name, plugin.name());
+            }
+        });
+    }
+
+    @Override
+    public PluginDescriptor find(final String organisation, final String name, final String version) {
+        // List plugins with same organisation (if known), name and version
+        Collection<PluginDescriptor> plugins = list(new Filter() {
+            @Override
+            public boolean accept(PluginDescriptor plugin) {
+                boolean match = true;
+                if ((organisation != null) && (plugin.organisation() != null)) {
+                    match = StringUtils.equalsIgnoreCase(organisation, plugin.organisation());
+                }
+                return match
+                        && StringUtils.equalsIgnoreCase(name, plugin.name())
+                        && StringUtils.equalsIgnoreCase(version, plugin.version());
+            }
+        });
+
+        // The same plugin can't exist multiple times... if so, reports the error
+        // because a manual clean up will be necessary
+        if ((plugins != null) && (!plugins.isEmpty())) {
```

Comment:
BTW I don't think it can be `null`. So testing `isEmpty()` might be enough here.
I'd write something like:

```java
if (plugins.size() > 1) {
   // throw ....
}
```